imports/wasi_snapshot_preview1/fs.go

```diff
@@ -707,14 +710,19 @@ func fdReaddirFn(ctx context.Context, mod api.Module, stack []uint64) (uint32, E
 	cookie := int64(stack[3])
 
 	// Validate the FD is a directory
-	rd, dir, errno := openedDir(ctx, mod, fd)
+	rd, dir, errno := openedDir(ctx, fsc, fd)
 	if errno != ErrnoSuccess {
 		return 0, errno
 	}
 
 	// expect a cookie only if we are continuing a read.
-	if cookie == 0 && dir.CountRead > 0 {
-		return 0, ErrnoInval // invalid as a cookie is minimally one.
+	if cookie == 0 {
+		// TODO: When bufLen < 8 we can't ever pass a cookie > 0 later, as the
```

Comment:
I think it makes no sense to support partial structs write. The given bufLen should always give at least room for 1 struct (without the filename), so 24 bytes. If the room left is less than 24 bytes I don't think it should write away another struct (that's also what maxDirents does).

So in my opinion, bufLen < 24 = ErrnoInval

Reply:
sounds great to me!